Feature/lockfiles new (#1790)
conan-io/docs: versioning/lockfiles/configurations.rst

.. _versioning_lockfiles_configurations:

Multiple configurations
=======================

.. warning::

    This is an **experimental** feature subject to breaking changes in future releases.

In the previous section we managed just 1 configuration, for the default profile. In many applications,
packages needs to be built with several different configurations, typically managed by different profile
files.

.. note::

    This section continues with the previous example with the :ref:`versioning_lockfiles_introduction`.
    The code used in this section, including a *build.py* script to reproduce it, is in the
    examples repository: https://github.com/conan-io/examples. You can go step by step
    reproducing this example while reading the below documentation.

    .. code:: bash

        $ git clone https://github.com/conan-io/examples.git
        $ cd features/lockfiles/intro
        # $ python build.py only to run the full example, but better go step by step

Lets start in the *features/lockfiles/intro* of the examples repository, remove the previous packages,
and create both release and debug ``pkga`` packages:

.. code-block:: bash

>     $ conan remove "*" -f

Reviewer: Maybe we can do `conan remove "pkg*" -f` so that the user that is following the tutorial does not delete the entire cache accidentally?

Author: Done

    $ conan create pkga pkga/0.1@user/testing
    $ conan create pkga pkga/0.1@user/testing -s build_type=Debug

Now, we could (don't do it) create 2 different lockfiles, one for each configuration:

.. code-block:: bash

    # DO NOT type these commands, we'll do it better below
    $ cd pkgb
    $ conan lock create conanfile.py --user=user --channel=testing --lockfile-out=locks/pkgb_release.lock
    $ conan lock create conanfile.py --user=user --channel=testing --lockfile-out=locks/pkgb_debug.lock -s build_type=Debug
        

.. important::

    The dependency graph is different for each different configuration/profile. Not only the package-ids, but also because of
    conditional requirements, the dependencies can be different. Then, it is necessary to create a lockfile for every different 
    configuration/profile. 

> But what if between both commands, the new ``pkga/0.2@user/testing`` version was created? Although this is unlikely in this

Reviewer 2: ```suggestion
But, what if a new ``pkga/0.2@user/testing`` version was created in the time between both commands,? Although this is unlikely to happen in this
```

> example, because everything is local, it is a possibility if ``pkga`` was in a server, and someone (or CI), upload the new

Reviewer 2: Let's encourage the CI usage

```suggestion
example, because everything is local. However, it could happen that ``pkga`` was in a server and the CI, uploads the new
```

version while we are running the above commands.

Base lockfiles
--------------

> Conan proposes a "base" lockfile, defined by the ``--base`` argument, that will capture only the versions and topology of the

Reviewer 2: ```suggestion
Conan proposes a "base" lockfile, with the :command:`--base` argument, that will capture only the versions and topology of the
```

    $ conan install .. --lockfile=../locks/pkgb_deps_debug.lock -s build_type=Debug
    ERROR: Cannot use profile, settings, options or env 'host' when using lockfile

results in an error, because the *locks/pkgb_deps_debug.lock* already stores the ``settings.build_type`` and passing it in the command line
could only result in inconsistencies and errors.

.. important::

    Lockfiles store the full effective profile configuration. It is not possible to pass configuration, settings, options or 
    profile arguments when using lockfiles (only when creating the lockfiles)

With the two captured lockfiles, now we can locally build and run our ``pkgb`` application for both configurations, guaranteeing
the dependency to ``pkga/0.1@user/testing``:

.. code-block:: bash

    $ conan install .. --lockfile=../locks/pkgb_deps_release.lock
    $ cmake ../src -G "Visual Studio 15 Win64"
    $ cmake --build . --config Release
    $ ./bin/greet
    HelloA 0.1 Release
    HelloB Release!
    Greetings Release!
    $ conan install .. --lockfile=../locks/pkgb_deps_debug.lock
    $ cmake --build . --config Debug
    $ ./bin/greet
    HelloA 0.1 Debug
    HelloB Debug!
    Greetings Debug!

> We can again create the ``pkgb`` package for both configurations:

Reviewer 2: ```suggestion
We can create ``pkgb`` package again for both configurations:
```

.. code-block:: bash

    $ cd ..
    $ conan create . user/testing --lockfile=locks/pkgb_deps_release.lock --lockfile-out=locks/pkgb_release.lock
    $ conan create . user/testing --lockfile=locks/pkgb_deps_debug.lock --lockfile-out=locks/pkgb_debug.lock

And we could still use the lockfiles later in time to install the ``pkgb`` package with the same dependencies
and configuration that were used to create that package:

.. code-block:: bash

    $ cd ..
    $ mkdir consume
    $ cd consume
    $ conan install pkgb/0.1@user/testing --lockfile=../pkgb/locks/pkgb_release.lock
    $ ./bin/greet
    HelloA 0.1 Release
    HelloB Release!
    Greetings Release!
    $ conan install pkgb/0.1@user/testing --lockfile=../pkgb/locks/pkgb_debug.lock
    $ ./bin/greet
    HelloA 0.1 Debug
    HelloB Debug!
    Greetings Debug!

> The immutability principle still holds. If we try to use *pkgb_release.lock* to create the ``pkgb`` package
> again instead of the *pkgb_deps_release.lock* lockfile, it will error, as ``pkgb`` would be already fully
> locked in the former.

Reviewer 2: ```suggestion
As you can see, the immutability principle remains. If we try to use *pkgb_release.lock* to create the ``pkgb`` package
again instead of the *pkgb_deps_release.lock* lockfile, it will error, as ``pkgb`` would be already fully
locked in the former.
```
